### PR TITLE
Estimates for select query

### LINKED ISSUE #23941: show estimates (estimated query complexity / parts / marks / rows)
Clickhouse knows before a query execution 
```
Selected 4/4 parts by partition key, 4 parts by primary key, 
126/126 marks by primary key, 126 marks to read from 4 ranges
Reading approx. 1032192 rows with 7 streams
```

I need ability to get these numbers without query execution:

```
show estimates for select count() from table;

┌─parts─┬─marks─┬────rows─┐
│   343 │  1126 │ 1032192 │
└───────┴───────┴─────────┘
```

also it seems they will be helpful in system.processes & system.query_log

### PR DESCRIPTION
**Pull request by fastio**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`EXPLAIN` query now has `EXPLAIN ESTIMATE ...` mode that will show information about read rows, marks and parts from MergeTree tables. Closes #23941.

Detailed description / Documentation draft:

Use case:

```
EXPLAIN ESTIMATES
SELECT
    toYear(LO_ORDERDATE) AS year,
    S_CITY,
    P_BRAND,
    sum(LO_REVENUE - LO_SUPPLYCOST) AS profit
FROM lineorder_flat
WHERE (S_NATION = 'UNITED STATES') AND ((year = 1997) OR (year = 1998)) AND (P_CATEGORY = 'MFGR#14')
GROUP BY
    year,
    S_CITY,
    P_BRAND
ORDER BY
    year ASC,
    S_CITY ASC,
    P_BRAND ASC

Query id: 2ac0e792-6a4c-4c09-95f0-b4d4d9a630c5

┌─database─┬─table──────────┬─parts─┬─────rows─┬─marks─┐
│ default  │ lineorder_flat │    14 │ 14430068 │  1780 │
└──────────┴────────────────┴───────┴──────────┴───────┘
```

#24478

### CODE REVIEW

**Comment by fastio** [comment]
Hi, how can I find out which test case is fail ?.

**Comment by alexey-milovidov** [comment]
@fastio In fact it is all green. The only failed integration test (and only in one of three checks) is known to be flaky.

**Comment by alexey-milovidov** [comment]
I feel that EXPLAIN ESTIMATE will be better than EXPLAIN ESTIMATES.

Because there is EXPLAIN ANALYZE in other DBMS and ANALYZE is a verb; ESTIMATE is both noun and verb, and if we treat it as a verb it will be natural to say "please, estimate this query"... it feels more natural to me :)

I can rename after merge.

File: src/Interpreters/InterpreterExplainQuery.cpp [hunk 1]
```diff
@@ -313,11 +333,33 @@ BlockInputStreamPtr InterpreterExplainQuery::executeImpl()
...
 
-    if (single_line)
-        res_columns[0]->insertData(buf.str().data(), buf.str().size());
-    else
-        fillColumn(*res_columns[0], buf.str());
+        auto settings = checkAndGetSettings<QueryPlanSettings>(ast.getSettings());
+        QueryPlan plan;
+
+        // It should output the result even the format is Null, for example EXPLAIN ESTIMATES select * from x format Null;
+        if (ast.format && ast.format->getColumnName() == "Null")
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterExplainQuery.cpp:345]: I found it controversial.
User may want to output result of EXPLAIN in any format and I would like to not make an exception for format Null.

**Review by alexey-milovidov — COMMENTED** [review]
LGTM, only a few changes needed.

**Comment by alexey-milovidov** [comment]
I will do the changes and merge.

**Comment by alexey-milovidov** [comment]
I cannot get correct results in test queries.

```
EXPLAIN ESTIMATE
SELECT count()
FROM test.hits
WHERE CounterID = 29103473

Query id: 4eda295e-6ee5-4ed6-a283-aacf38572717

┌─database─┬─table─┬─parts─┬────rows─┬─marks─┬──────bytes─┐
│ test     │ hits  │     1 │ 8873898 │  1084 │ 1271407349 │
└──────────┴───────┴───────┴─────────┴───────┴────────────┘

1 rows in set. Elapsed: 0.003 sec. 

milovidov-desktop :) SELECT count() FROM test.hits WHERE CounterID = 29103473

SELECT count()
FROM test.hits
WHERE CounterID = 29103473

Query id: e04582bf-108f-4524-b4c0-f275673cdfc5

┌─count()─┐
│   47758 │
└─────────┘

1 rows in set. Elapsed: 0.003 sec. Processed 57.34 thousand rows, 229.38 KB (20.33 million rows/s., 81.30 MB/s.)
```

Explained amount of data is too large (both rows and bytes).

**Comment by alexey-milovidov** [comment]
Maybe @KochetovNicolai may help?

**Comment by fastio** [comment]
> I cannot get correct results in test queries.
> 
> ```
> EXPLAIN ESTIMATE
> SELECT count()
> FROM test.hits
> WHERE CounterID = 29103473
> 
> Query id: 4eda295e-6ee5-4ed6-a283-aacf38572717
> 
> ┌─database─┬─table─┬─parts─┬────rows─┬─marks─┬──────bytes─┐
> │ test     │ hits  │     1 │ 8873898 │  1084 │ 1271407349 │
> └──────────┴───────┴───────┴─────────┴───────┴────────────┘
> 
> 1 rows in set. Elapsed: 0.003 sec. 
> 
> milovidov-desktop :) SELECT count() FROM test.hits WHERE CounterID = 29103473
> 
> SELECT count()
> FROM test.hits
> WHERE CounterID = 29103473
> 
> Query id: e04582bf-108f-4524-b4c0-f275673cdfc5
> 
> ┌─count()─┐
> │   47758 │
> └─────────┘
> 
> 1 rows in set. Elapsed: 0.003 sec. Processed 57.34 thousand rows, 229.38 KB (20.33 million rows/s., 81.30 MB/s.)
> ```
> 
> Explained amount of data is too large (both rows and bytes).

hi @alexey-milovidov  Thanks for your review. I will fix the problems soon.

**Comment by fastio** [comment]
If want to obtain the exact amount of bytes to read, we should create the marks loader object during the building query pipeline stage. So I remove the bytes column.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 2]
```diff
@@ -838,13 +841,17 @@ ReadFromMergeTree::AnalysisResult ReadFromMergeTree::selectRangesToRead(MergeTre
 
     size_t sum_marks = 0;
     size_t sum_ranges = 0;
+    size_t sum_rows = 0;
 
     for (const auto & part : result.parts_with_ranges)
     {
         sum_ranges += part.ranges.size();
         sum_marks += part.getMarksCount();
+        sum_rows += part.getRowsCount();
```
Comment (alexey-milovidov) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:850]: But we can read subset of rows and marks by using primary key or secondary indices.

Comment (fastio) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp:850]: True, the `result.parts_with_ranges is returned by ReadFromMergeTree::filterPartsByPrimaryKeyAndSkipIndexes function.  the variable part is object of RangesInDataPart. So the count of marks and rows is what we want.

Comment (fastio) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp:850]: > But we can read subset of rows and marks by using primary key or secondary indices.

True, the `result.parts_with_ranges is returned by ReadFromMergeTree::filterPartsByPrimaryKeyAndSkipIndexes function. the variable part is object of RangesInDataPart. So the count of marks and rows is what we want.

**Comment by alexey-milovidov** [comment]
Yes, it works :)

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@fastio You have somehow deleted my test.

**Comment by fastio** [comment]
> @fastio You have somehow deleted my test.

Sorry, I force pushed the commit. I recover it.

**Comment by alexey-milovidov** [comment]
```
diff --git a/tests/queries/1_stateful/00166_explain_estimate.reference b/tests/queries/1_stateful/00166_explain_estimate.reference
new file mode 100644
index 0000000000..71ddd68158
--- /dev/null
+++ b/tests/queries/1_stateful/00166_explain_estimate.reference
@@ -0,0 +1,5 @@
+test   hits    1       57344   7
+test   hits    1       8839168 1079
+test   hits    1       835584  102
+test   hits    1       8003584 977
+test   hits    2       581632  71
diff --git a/tests/queries/1_stateful/00166_explain_estimate.sql b/tests/queries/1_stateful/00166_explain_estimate.sql
new file mode 100644
index 0000000000..06725ff7f9
--- /dev/null
+++ b/tests/queries/1_stateful/00166_explain_estimate.sql
@@ -0,0 +1,5 @@
+EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID = 29103473;
+EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID != 29103473;
+EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID > 29103473;
+EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID < 29103473;
+EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID = 29103473 UNION ALL SELECT count() FROM test.hits WHERE CounterID = 1704509;
```

**Comment by fastio** [comment]
> @fastio You have somehow deleted my test.



> ```
> diff --git a/tests/queries/1_stateful/00166_explain_estimate.reference b/tests/queries/1_stateful/00166_explain_estimate.reference
> new file mode 100644
> index 0000000000..71ddd68158
> --- /dev/null
> +++ b/tests/queries/1_stateful/00166_explain_estimate.reference
> @@ -0,0 +1,5 @@
> +test   hits    1       57344   7
> +test   hits    1       8839168 1079
> +test   hits    1       835584  102
> +test   hits    1       8003584 977
> +test   hits    2       581632  71
> diff --git a/tests/queries/1_stateful/00166_explain_estimate.sql b/tests/queries/1_stateful/00166_explain_estimate.sql
> new file mode 100644
> index 0000000000..06725ff7f9
> --- /dev/null
> +++ b/tests/queries/1_stateful/00166_explain_estimate.sql
> @@ -0,0 +1,5 @@
> +EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID = 29103473;
> +EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID != 29103473;
> +EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID > 29103473;
> +EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID < 29103473;
> +EXPLAIN ESTIMATE SELECT count() FROM test.hits WHERE CounterID = 29103473 UNION ALL SELECT count() FROM test.hits WHERE CounterID = 1704509;
> ```

Thanks :)

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
The test should be in skip_list.json for DatabaseReplicated, because in DatabaseReplicated data is split in different set of parts.

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-11968

**Pull request merged**